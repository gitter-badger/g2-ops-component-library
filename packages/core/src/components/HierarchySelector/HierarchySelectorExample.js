import React from 'react'
import renderIf from 'render-if'
import { prop } from 'ramda'
import { formatDate, daysElapsedUntilToday } from '../DatePicker/dateUtils'
import { FlattenedOptionType, OptionType } from 'types/HierarchySelector'
import {HierarchySelector} from './HierarchySelector'
import towProvidersJson from './towProviders.json'

// LEHOFSTADT - DEFAULT DISPATCH - GROUP - DEVIN BURT
const initialUser = {
  company: 42517,
  group: 40348,
  driver: 15235,
}

const initialOptionId = prop('driver', initialUser) || prop('group', initialUser) || prop('company', initialUser)

export const CardIcon = () => (
  <div className="CardIcon Icon" data-tip="PCard">
    <div className="red circle" />
    <div className="orange circle" />
  </div>
)

export const renderMethod = (option) => {
  const renderIfCompany = renderIf(option.level === 'company' && option.expiresOn)
  const labelColor = option.dispatch_flag === true ? 'black' : 'grey'
  const expiryDateColor = option.isExpired ? 'red' : 'black'
  const renderIfCard = renderIf(option.p_card_flag === 'Y')
  return (
    <div className="hierarchyOption">
      <span style={{ color: `${labelColor}` }}>{option.label}</span>
      {renderIfCompany(
        <span className="separator">
          <i>
            Expires: <span style={{ color: `${expiryDateColor}` }}>{option.expiresOn} </span>
          </i>
        </span>,
      )}
      {renderIfCard(
        <span className="separator">
          <CardIcon />
        </span>,
      )}
    </div>
  )
}
const isExceeded = (givenDate: string, numberOfDays: number) => daysElapsedUntilToday(givenDate) > numberOfDays

function transformPersonnelOption(option, isVendorExpired) {
  const componentOption: OptionType = {
    id: option.vendor_personnel_id,
    level: 'driver',
    label: `${option.first_name} ${option.last_name}`,
    dispatch_flag: option.dispatch_flag,
    p_card_flag: option.p_card_flag,
    isExpired: isVendorExpired,
    isSelectable: !isVendorExpired && option.dispatch_flag === true,
  }
  return componentOption
}

function transformGroupOption(option, isVendorExpired) {
  const componentOption: OptionType = {
    id: option.dispatch_group_id,
    level: 'group',
    label: option.dispatch_group_name,
    dispatch_flag: option.dispatch_flag,
    p_card_flag: option.p_card_flag,
    isExpired: isVendorExpired,
    isSelectable: !isVendorExpired && option.dispatch_flag === true,
    options: option.personnel.map((personnelOption) => transformPersonnelOption(personnelOption, isVendorExpired)),
  }
  return componentOption
}

function transformVendorOption(option) {
  const formattedDate: OptionType =
    option.insurance_expiration_date && formatDate(option.insurance_expiration_date, 'YYYY-MM-DDTHH:mm:sss Z')
  const isVendorExpired = formattedDate && isExceeded(formattedDate, 30)
  const componentOption = {
    id: option.vendor_id,
    level: 'company',
    label: option.vendor_short_name,
    dispatch_flag: option.dispatch_flag,
    expiresOn: formattedDate,
    p_card_flag: option.p_card_flag,
    isExpired: isVendorExpired,
    isSelectable: !isVendorExpired && option.dispatch_flag === true,
    options: option.dispatch_groups.map((groupOption) => transformGroupOption(groupOption, isVendorExpired)),
  }
  return componentOption
}
const towProviders = towProvidersJson.data
const componentOptions = towProviders.map(transformVendorOption)

class Example extends React.Component {
  state = {
    value: null,
  }
  onChange = (selectedOption: FlattenedOptionType) => {
    const selectedHierarchyOption = (typeof selectedOption === 'string')? selectedOption : selectedOption.hierarchy
    this.setState({ value: selectedHierarchyOption })
  }
  render () {
    return (
      <div style={{ maxWidth: '400px' }}>
        <HierarchySelector
          name="Tow Provider"
          label="Tow Provider"
          options={componentOptions}
          width={200}
          optionStyleProps={{ rowHeight: 40, optionsMinHeight: 200, fontSize: '12px' }}
          value={initialUser}
          renderMethod={renderMethod}
          onChange={this.onChange}
          onRenderSuffix={() => <i className="material-icons">arrow_drop_down</i>}
        />
      </div>
    )
  }
}

export default Example
