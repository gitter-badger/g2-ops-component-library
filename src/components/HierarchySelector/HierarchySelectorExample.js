import React from 'react'
import renderIf from 'render-if'

import { formatDate, daysElapsedUntilToday } from '../DatePicker/dateUtils'
import HierarchySelector from './HierarchySelector'
import './HierarchySelector.scss'
import { FlattenedOptionType, HierarchyType } from '../../../types/HierarchySelector'
import towProvidersJson from './towProviders.json'

// G/RTOW - G/RTOWDG - ALLEN MARQUEZ
const initialUser = {
  company: 226,
  group: 203,
  driver: 1000,
}

const handleSelectSubhauler = (selectedOption: FlattenedOptionType) => {
  const selectedSubhauler = { user: selectedOption.hierarchy }
  console.log('handleSelectSubhauler called for newly selected value: ', selectedSubhauler)
}
export const CardIcon = () => (
  <div className="CardIcon Icon" data-tip="PCard">
    <div className="red circle" />
    <div className="orange circle" />
  </div>
)

export const renderMethod = (option, isExpired) => {
  const renderIfCompany = renderIf(option.level === 'company' && option.expiresOn)
  const labelColor = option.isDisabled ? 'grey' : 'black'
  const expiryDateColor = isExpired ? 'red' : 'black'
  const renderIfCard = renderIf(option.p_card_flag === 'Y')
  return (
    <div className="hierarchyOption">
      <span style={{ color: `${labelColor}` }}>{option.label}</span>
      {renderIfCompany(
        <span className="separator">
          <i>
            Expires: <span style={{ color: `${expiryDateColor}` }}>{option.expiresOn} </span>
          </i>
        </span>,
      )}
      {renderIfCard(
        <span className="separator">
          <CardIcon />
        </span>,
      )}
    </div>
  )
}
const isExceeded = (givenDate: string, numberOfDays: number) => daysElapsedUntilToday(givenDate) > numberOfDays

function transformPersonnelOption(option) {
  const componentOption = {
    id: option.vendor_personnel_id,
    level: 'driver',
    label: `${option.first_name} ${option.last_name}`,
    isDisabled: option.dispatch_flag === false || false,
    p_card_flag: option.p_card_flag,
  }
  return componentOption
}

function transformGroupOption(option) {
  const componentOption = {
    id: option.dispatch_group_id,
    level: 'group',
    label: option.dispatch_group_name,
    isDisabled: option.dispatch_flag === false || false,
    p_card_flag: option.p_card_flag,
    options: option.personnel.map(transformPersonnelOption),
  }
  return componentOption
}

function transformVendorOption(option) {
  const formattedDate =
    option.insurance_expiration_date && formatDate(option.insurance_expiration_date, 'YYYY-MM-DDTHH:mm:sss Z')
  const isVendorExpired = formattedDate && isExceeded(formattedDate, 30)
  const componentOption = {
    id: option.vendor_id,
    level: 'company',
    label: option.vendor_short_name,
    expiresOn: formattedDate,
    isExpired: !!isVendorExpired,
    isDisabled: option.dispatch_flag === false || false,
    p_card_flag: option.p_card_flag,
    options: option.dispatch_groups.map(transformGroupOption),
  }
  return componentOption
}
const towProviders = towProvidersJson.data
const componentOptions = towProviders.map(transformVendorOption)

console.log('component: ', componentOptions)
const Example = () => (
  <div style={{ maxWidth: '400px' }}>
    <HierarchySelector
      name="HierarchySelector Field"
      options={componentOptions}
      width={200}
      optionStyleProps={{ rowHeight: 40, optionsMinHeight: 200, fontSize: '12px' }}
      value={initialUser}
      renderMethod={renderMethod}
      onChange={handleSelectSubhauler}
    />
  </div>
)

export default Example
