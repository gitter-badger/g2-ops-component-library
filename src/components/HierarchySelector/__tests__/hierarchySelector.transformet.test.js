// @flow

import { flattenNestedOptions } from '../hierarchySelector.transformer'
import { renderMethod } from '../HierarchySelectorExample'

describe('Handle Transformation', () => {
  test('properly transform nested option', () => {
    const nestedOptions = [
      {
        name: 'company1',
        label: "Cecy's Towing Company",
        level: 0,
        expiresOn: '04/15/2018',
        options: [
          {
            name: 'group1',
            label: "Cecy's Group",
            level: 1,
            options: [
              {
                name: 'driver1',
                label: 'Manuel R Cero fgfg jhjh kjkj hjhj hjh jhjhj jh jh jhj j j hj jh jh j hj hj j jh jh jh j hj4',
                level: 2,
                isDispatchable: true,
                hasPcard: true,
              },
              {
                name: 'driver2',
                label: 'Driver 2',
                level: 2,
              },
              {
                name: 'driver2',
                label: 'Driver 3',
                level: 2,
                hasPcard: true,
              },
            ],
          },
          {
            name: 'group2',
            label: "Cecy's Group2",
            isDisabled: true,
            level: 1,
            options: [
              {
                name: 'driver1',
                label: 'Marc Jacobs',
                level: 2,
              },
            ],
          },
        ],
      },
      {
        name: 'company2',
        label: 'Haul O Way',
        expiresOn: '04/15/2018',
        level: 0,
        options: [
          {
            name: 'group1',
            label: "Haul O' Way",
            level: 1,
            options: [
              {
                name: 'driver1',
                label: 'Jason',
                level: 2,
              },
            ],
          },
          {
            name: 'group2',
            label: "Haul O' Way Group",
            level: 1,
            options: [
              {
                name: 'driver2',
                label: 'Anderson',
                level: 2,
                isDisabled: true,
              },
            ],
          },
        ],
      },
      {
        name: 'company3',
        label: 'ABC Company',
        expiresOn: '04/15/2018',
        level: 0,
        options: [
          {
            name: 'group1',
            label: 'ABC Group',
            level: 1,
            options: [
              {
                name: 'driver1',
                label: 'Jason',
                level: 2,
              },
            ],
          },
        ],
      },
      {
        name: 'company4',
        label: 'XYZ Company',
        expiresOn: '01/15/2018',
        level: 0,
        options: [
          {
            name: 'group1',
            label: 'XYZ Group',
            level: 1,
            options: [
              {
                name: 'driver1',
                label: 'Jason',
                level: 2,
              },
            ],
          },
        ],
      },
    ]
    expect(flattenNestedOptions(nestedOptions, renderMethod)).toMatchSnapshot()
  })
  test('handle empty options', () => {
    const nestedOption = [
      {
        name: 'company1',
        label: 'company1',
        options: [],
      },
    ]
    expect(flattenNestedOptions(nestedOption, renderMethod)).toMatchSnapshot()
  })
})
