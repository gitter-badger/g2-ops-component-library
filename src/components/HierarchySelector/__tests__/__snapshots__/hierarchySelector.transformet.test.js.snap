// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handle Transformation handle empty options 1`] = `
Array [
  Object {
    "haveChildren": false,
    "label": "Asia",
    "name": "asia",
    "path": Array [
      "Asia",
    ],
  },
]
`;

exports[`Handle Transformation properly transform nested option 1`] = `
Array [
  Object {
    "haveChildren": true,
    "label": "Asia",
    "name": "asia",
    "path": Array [
      "Asia",
    ],
  },
  Object {
    "haveChildren": true,
    "label": "India",
    "name": "india",
    "path": Array [
      "Asia",
      "India",
    ],
  },
  Object {
    "haveChildren": false,
    "label": "Telangana",
    "name": "telangana",
    "path": Array [
      "Asia",
      "India",
      "Telangana",
    ],
  },
  Object {
    "haveChildren": true,
    "label": "China",
    "name": "china",
    "path": Array [
      "Asia",
      "China",
    ],
  },
  Object {
    "haveChildren": false,
    "label": "Beijing",
    "name": "beijing",
    "path": Array [
      "Asia",
      "China",
      "Beijing",
    ],
  },
  Object {
    "haveChildren": true,
    "label": "North America",
    "name": "northAmerica",
    "path": Array [
      "North America",
    ],
  },
  Object {
    "haveChildren": true,
    "label": "USA",
    "name": "usa",
    "path": Array [
      "North America",
      "USA",
    ],
  },
  Object {
    "haveChildren": false,
    "label": "Texas",
    "name": "texas",
    "path": Array [
      "North America",
      "USA",
      "Texas",
    ],
  },
  Object {
    "haveChildren": false,
    "label": "Arkansas",
    "name": "arkansas",
    "path": Array [
      "North America",
      "USA",
      "Arkansas",
    ],
  },
  Object {
    "haveChildren": true,
    "label": "Africa",
    "name": "africa",
    "path": Array [
      "Africa",
    ],
  },
  Object {
    "haveChildren": false,
    "label": "Egypt",
    "name": "egypt",
    "path": Array [
      "Africa",
      "Egypt",
    ],
  },
  Object {
    "haveChildren": false,
    "label": "Libya",
    "name": "libya",
    "path": Array [
      "Africa",
      "Libya",
    ],
  },
]
`;
